tokio/src/io/driver/mod.rs

```diff
@@ -181,12 +194,22 @@ impl Driver {
 
 impl Drop for Driver {
     fn drop(&mut self) {
-        self.resources.for_each(|io| {
-            // If a task is waiting on the I/O resource, notify it. The task
-            // will then attempt to use the I/O resource and fail due to the
-            // driver being shutdown.
-            io.wake(Ready::ALL);
-        })
+        (*self.inner.resources.lock()) = self.resources.take();
+    }
+}
+
+impl Drop for Inner {
+    fn drop(&mut self) {
+        let resources = self.resources.lock().take();
+
+        if let Some(mut slab) = resources {
+            slab.for_each(|io| {
+                // If a task is waiting on the I/O resource, notify it. The task
+                // will then attempt to use the I/O resource and fail due to the
+                // driver being shutdown.
+                io.shutdown();
+            });
+        }
```

Comment:
I think that changes things a bit. Before the idea was that the moment the driver gets dropped, it wakes all io tasks. Now however, there might be strong arc(s) being held somewhere else. So this logic is being delayed until all these arcs are dropped. Is that desirable? Does that actually solve the original racing  issue? Might be missing something here...

Reply:
It should not change things materially. The strong refs are only held for small amounts of time to perform registration / deregistration.

Reply:
It's possible to keep the Inner alive with enough traffic on the handle, but this isn't worse than it was before (it's better, since eventually once all the Arc cloning stops and the Inner can be dropped, all the new handles that were registered after the driver was dropped will be properly notified)